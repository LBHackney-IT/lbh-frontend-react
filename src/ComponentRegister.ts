import { ElementType, ReactNode } from "react";

/**
 * The property types for {@link LinkComponent}.
 *
 * @property {string} [className]
 * @property {string} href
 * @property {ReactNode} children
 */
export interface LinkComponentProps {
  className?: string;
  href: string;
  children: ReactNode;
}

/**
 * The type of a valid link component to be passed to {@link ComponentRegister}.
 * It allows any valid React component (including tag strings) that accept
 * {@link LinkComponentProps} as proptypes.
 */
export type LinkComponent = ElementType<LinkComponentProps>;

interface ComponentRegisterDefaultOptions {
  components: {
    Link: LinkComponent;
  };
}

/**
 * The default options for {@link ComponentRegister}.
 */
export const defaultComponentOptions: ComponentRegisterDefaultOptions = {
  components: {
    Link: "a"
  }
};

/**
 * The properties for registering components.
 *
 * @property {object} [components] - The components to register
 * @property {LinkComponent} [components.Link] - The component to use internally
 * for handling links. Defaults to `"a"`.
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface ComponentRegisterOptions
  extends Partial<ComponentRegisterDefaultOptions> {}

/**
 * A register of components to use with library components with sensible
 * defaults. Most users will never need to interact with this. It exists to
 * support customization of library component behaviour.
 */
export class ComponentRegister {
  /**
   * @static The {@link LinkComponent} to allow `lbh-frontend-react` to hook
   * into routers or other custom linking libraries.
   *
   * Defaults to `"a"`. See {@link defaultComponentOptions}.
   */
  static Link: LinkComponent = defaultComponentOptions.components.Link;

  /**
   * @static Resets the component register to its defaults. See
   * {@link defaultComponentOptions} for those defaults.
   */
  static reset(): void {
    this.init(defaultComponentOptions);
  }

  /**
   * @static Initializes {@link ComponentRegister} with the provided options.
   *
   * This is how you set component options en masse, and is the preferred method
   * for setting custom components.
   *
   * @example
   * // To set `lbh-frontend-react` up to use React Router's `Link` component.
   * ComponentRegister.init({
   *   components.
   *   Link: ({ className, href, children }) => (
   *     <ReactRouter.Link className={className} to={href}>
   *       {children}
   *     </ReactRouter.Link>
   *   )
   * });
   *
   * @param {ComponentRegisterOptions} [options] - Options to customize the
   * underlying components used by library components.
   */
  static init({ components }: ComponentRegisterOptions = {}): void {
    if (components && components.Link) {
      this.Link = components.Link;
    }
  }
}
