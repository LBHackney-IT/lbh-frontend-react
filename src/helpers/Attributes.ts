import React from "react";

/**
 * The proptypes to support `data-*` attributes.
 */
export interface DataAttributes {
  /**
   * Any `data-*` attributes.
   */
  // This tells TypeScript / JSX to also accept any `data-*`  attributes when
  // constructing the component. Like magic!
  data?: string;
}

/**
 * A helper for parsing wildcard attributes.
 */
export class Attributes {
  /**
   * Select all `aria-*` attributes from props.
   */
  static aria<Props extends React.AriaAttributes>(
    props: Props
  ): React.AriaAttributes {
    return Object.entries(props)
      .filter(([key]) => key.startsWith("aria-"))
      .reduce(
        (attributes, [key, value]) => ({ ...attributes, [key]: value }),
        {} as React.AriaAttributes
      );
  }

  /**
   * Select all `data-*` attributes from props.
   */
  static data<Props extends DataAttributes>(props: Props): DataAttributes {
    return Object.entries(props)
      .filter(([key]) => key.startsWith("data-"))
      .reduce(
        (attributes, [key, value]) => ({ ...attributes, [key]: value }),
        {} as DataAttributes
      );
  }

  /**
   * Select all `aria-*` and `data-*` attributes from props.
   */
  static ariaAndData<Props extends React.AriaAttributes & DataAttributes>(
    props: Props
  ): React.AriaAttributes & DataAttributes {
    return {
      ...this.aria(props),
      ...this.data(props)
    };
  }
}
