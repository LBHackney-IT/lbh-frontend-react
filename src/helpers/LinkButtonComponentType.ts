/**
 * The proptypes for {@link LinkButtonComponentType}.
 *
 * There is an additional required prop missing from this interface,
 * `onKeyDown`, which follows the usual React interface
 * `React.KeyboardEventHandler`.
 */
export interface LinkButtonComponentTypeProps {
  id?: string;
  className?: string;
  /**
   * Use for the main call to action on your service's start page. Adjusts
   * style of button and adds a start icon.
   */
  isStartButton?: boolean;
  href: string;
  children: React.ReactNode;
}

/**
 * The type of a valid link component to be passed to {@link ComponentRegister}
 * to be used for {@link LinkButton}. It allows any valid React component
 * (including tag strings) that accept {@link LinkButtonComponentTypeProps} as
 * proptypes.
 */
export type LinkButtonComponentType = React.ElementType<
  LinkButtonComponentTypeProps
>;
