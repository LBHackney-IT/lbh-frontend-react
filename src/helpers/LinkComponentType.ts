/**
 * The proptypes for {@link LinkComponentType}.
 */
export interface LinkComponentTypeProps {
  id?: string;
  className?: string;
  href: string;
  children: React.ReactNode;
}

/**
 * The type of a link component to be passed to {@link ComponentRegister}.
 * It allows any valid React component (including tag strings) that accept
 * {@link LinkComponentTypeProps} as proptypes.
 */
export type LinkComponentType = React.ElementType<LinkComponentTypeProps>;
