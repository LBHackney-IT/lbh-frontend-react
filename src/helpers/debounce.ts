/**
 * The context to pass to {@link debounce}.
 *
 * Create a single instance of this for any given component via
 * {@link makeDebounceContext}. Persist it between calls of {@link debounce} for
 * the same element.
 */
export interface DebounceContext {
  /**
   * @ignore
   */
  shouldDebounce: boolean;
}

/**
 * Creates a new instance of {@link DebounceContext}.
 */
export const makeDebounceContext = (): DebounceContext => ({
  shouldDebounce: false,
});

/**
 * A function used to debounce clicks occuring close together (that is, double
 * clicks) if the attribute `data-prevent-double-click` is set to `"true"`.
 *
 * If debouncing is appropriate, this function calls `event.preventDefault()`
 * automatically.
 *
 * @param event - The click event to possible debounce.
 *
 * @param context - An instance of {@link DebounceContext}, persisted between
 * debounce calls for the same element.
 *
 * @param debounceTimeout - The time between click to consider a double click in
 * milliseconds. Defaults to `1000`.
 *
 * @returns `true` if the click should be debounced, otherwise `false`.
 */
// This is a reimplementation of GOV.UK Frontend's `Button.protoype.debounce`,
// sourced from version `3.3.0`. To understand why, see `/docs/adr/0012-...`.
//
// See
// https://github.com/alphagov/govuk-frontend/blob/c473c7ec3123e220d83221b9fd12e216f71bda84/package/govuk/components/button/button.js
// for the source and
// https://github.com/alphagov/govuk-frontend/blob/c473c7ec3123e220d83221b9fd12e216f71bda84/LICENSE.txt
// for the licence applying to the original implementation.
export const debounce = (
  event: React.MouseEvent<HTMLElement>,
  context: DebounceContext,
  debounceTimeout = 1000
): boolean => {
  if (
    (event.target as HTMLElement).getAttribute("data-prevent-double-click") !==
    "true"
  ) {
    return false;
  }

  if (context.shouldDebounce) {
    event.preventDefault();

    return true;
  }

  const clearTimeout = (): void => {
    context.shouldDebounce = false;
  };

  context.shouldDebounce = true;

  setTimeout(clearTimeout, debounceTimeout);

  return false;
};
