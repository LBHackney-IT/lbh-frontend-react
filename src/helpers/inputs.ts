export const getInputId = (
  input: { id?: string; value: string },
  idPrefix: string
): string => {
  let id;
  if (input.id) {
    id = input.id;
  } else {
    id = `${idPrefix}-${input.value.replace(/\s+/g, "-").toLowerCase()}`;
  }
  return id;
};
