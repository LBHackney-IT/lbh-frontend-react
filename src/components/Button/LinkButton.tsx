import classNames from "classnames";
import GovukButton from "govuk-frontend/govuk/components/button/button";
import PropTypes from "prop-types";
import React, { Component } from "react";

import { nullAsUndefined } from "../../helpers/nullAsUndefined.internal";
import { ComponentRegister } from "../../helpers/ComponentRegister";
import { StartButtonIcon } from "../icons/StartButtonIcon";

import "lbh-frontend/lbh/components/lbh-button/_button.scss";

/**
 * The proptypes for {@link LinkButtonComponentType}.
 */
export interface LinkButtonProps {
  id?: string | null;
  className?: string | null;
  /**
   * Use for the main call to action on your service's start page. Adjusts
   * style of button and adds a start icon.
   */
  isStartButton?: boolean | null;
  href: string;
  children: React.ReactNode;
}

/**
 * A component for creating an anchor linking to other pages, styled as a
 * button. You can customize the component used internally for integration with
 * routers using {@link ComponentRegister.init}. By default, it uses an `<a>`
 * tag.
 */
export class LinkButton extends Component<LinkButtonProps> {
  static propTypes: PropTypes.ValidationMap<LinkButtonProps> = {
    id: PropTypes.string,
    className: PropTypes.string,
    isStartButton: PropTypes.bool,
    href: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired
  };

  /**
   * @ignore
   */
  render(): JSX.Element {
    const { id, className, isStartButton, href, children } = this.props;
    return (
      <ComponentRegister.LinkButton
        id={nullAsUndefined(id)}
        className={classNames("govuk-button lbh-button", className, {
          "govuk-button--start": isStartButton
        })}
        href={href}
        role="button"
        onKeyDown={GovukButton.prototype.handleKeyDown.bind(this)}
      >
        {children}
        {isStartButton && <StartButtonIcon />}
      </ComponentRegister.LinkButton>
    );
  }
}
