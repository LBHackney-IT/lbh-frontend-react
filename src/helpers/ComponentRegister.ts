import { LinkComponentType } from "./LinkComponentType";
import { LinkButtonComponentType } from "./LinkButtonComponentType";

/**
 * Components to register with {@link ComponentRegister} via
 * {@link ComponentRegisterOptions.components}.
 */
export interface ComponentRegisterComponents {
  /**
   * The component to use internally for handling links.
   *
   * Defaults to `"a"`.
   *
   */
  Link?: LinkComponentType;

  /**
   * The component to use internally for handling links. Similar to `Link`,
   * however is used to display links as buttons, rather than text.
   *
   * Defaults to `"a"`.
   */
  LinkButton?: LinkButtonComponentType;
}

/**
 * Options to pass to {@link ComponentRegister.init}.
 *
 * This is only generic to support sharing the type with
 * {@link defaultComponentRegisterOptions}. You should never need to override
 * the default generic parameter. It defaults to
 * {@link ComponentRegisterComponents}.
 */
export interface ComponentRegisterOptions<
  C extends ComponentRegisterComponents = ComponentRegisterComponents
> {
  /**
   * The components to register.
   *
   * Defaults to be of type {@link ComponentRegisterComponents}.
   */
  components?: C;
}

/**
 * The default options for {@link ComponentRegister}.
 */
export const defaultComponentRegisterOptions: Required<ComponentRegisterOptions<
  Required<ComponentRegisterComponents>
>> = {
  components: {
    Link: "a",
    LinkButton: "a"
  }
};

/**
 * A register of components to use within library components, with sensible
 * defaults. Many users will never need to interact with this. It exists to
 * support customization of library component behaviour.
 */
export class ComponentRegister {
  /**
   * The {@link LinkComponentType} to allow this library to hook into routers or
   * other custom linking libraries.
   *
   * Defaults to `"a"`. See {@link ComponentRegisterComponents.Link} and
   * {@link defaultComponentRegisterOptions}.
   */
  static Link: LinkComponentType =
    defaultComponentRegisterOptions.components.Link;

  /**
   * The {@link LinkButton} to allow this library to hook into routers or
   * other custom linking libraries. Similar to `Link`, however is displayed as
   * a button, rather than text.
   *
   * Defaults to `"a"`. See {@link ComponentRegisterComponents.LinkButton} and
   * {@link defaultComponentRegisterOptions}.
   */
  static LinkButton: LinkButtonComponentType =
    defaultComponentRegisterOptions.components.LinkButton;

  /**
   * Resets the component register to its defaults. See
   * {@link defaultComponentRegisterOptions} for those defaults.
   */
  static reset(): void {
    this.init(defaultComponentRegisterOptions);
  }

  /**
   * Initializes {@link ComponentRegister} with the provided options.
   *
   * This is how you set component options en masse, and is the preferred method
   * for providing custom components.
   *
   * If you need to change {@link ComponentRegisterComponents.Link}, you may
   * also need to change {@link ComponentRegisterComponents.LinkButton} given
   * that you wish for the functionality of both to be the same.
   *
   * ```ts
   * // Using `Link` from React Router.
   * ComponentRegister.init({
   *   components: {
   *     Link: ({ className, href, children }) => (
   *       <ReactRouter.Link className={className} to={href}>
   *         {children}
   *       </ReactRouter.Link>
   *     )
   *   }
   * });
   * ```
   *
   * ```ts
   * // Using `Link` from NextJS as `LinkButton`.
   * ComponentRegister.init({
   *   components: {
   *     LinkButton: ({id, className, href, children, onKeyDown }) => (
   *       <Link id={id} className={className} href={href} onKeyDown={onKeyDown}>
   *         {children}
   *       </Link>
   *     )
   *   }
   * });
   * ```
   *
   * @param options - Options to customize the underlying components used by
   * library components.
   */
  static init(options: ComponentRegisterOptions = {}): void {
    const { components } = options;

    if (components) {
      const { Link, LinkButton } = components;

      if (Link) {
        this.Link = Link;
      }

      if (LinkButton) {
        this.LinkButton = LinkButton;
      }
    }
  }
}
