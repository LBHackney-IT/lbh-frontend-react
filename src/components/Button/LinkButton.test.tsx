import GovukButton from "govuk-frontend/govuk/components/button/button";
import React from "react";
import { create } from "react-test-renderer";
import { mount } from "enzyme";

import { LinkButton } from "./LinkButton";
import {
  LinkButtonComponentType,
  LinkButtonComponentTypeProps
} from "../../helpers/LinkButtonComponentType";
import { ComponentRegister } from "../../helpers/ComponentRegister";

it("renders correctly with all props", () => {
  const component = create(
    <LinkButton id="1234" className="class1 class2" isStartButton href="/">
      Testchildren
    </LinkButton>
  );

  expect(component).toMatchSnapshot();
});

it("renders correctly without optional props", () => {
  const component = create(<LinkButton href="/">Testchildren</LinkButton>);

  expect(component).toMatchSnapshot();
});

it("clicks the button when the space key is pressed", () => {
  const mockClickHandler = jest.fn();

  const handleKeydownSpy = jest.spyOn(GovukButton.prototype, "handleKeyDown");

  const wrapper = mount(<LinkButton href="/">Testchildren</LinkButton>);

  const a = wrapper.find("a");

  a.getDOMNode().addEventListener("click", mockClickHandler);

  // 32 is the javascript event keycode for the space key
  a.simulate("keydown", { keyCode: 32 });

  expect(handleKeydownSpy).toHaveBeenCalledTimes(1);
  expect(mockClickHandler).toHaveBeenCalledTimes(1);
});

it("doesn't click the button when a key that isn't space is pressed ", () => {
  const mockClickHandler = jest.fn();

  const handleKeydownSpy = jest.spyOn(GovukButton.prototype, "handleKeyDown");

  const wrapper = mount(<LinkButton href="/">Testchildren</LinkButton>);

  const a = wrapper.find("a");

  a.getDOMNode().addEventListener("click", mockClickHandler);

  // 74 is the javascript event keycode for the "J" key
  a.simulate("keydown", { keyCode: 74 });

  expect(handleKeydownSpy).toHaveBeenCalledTimes(1);
  expect(mockClickHandler).toHaveBeenCalledTimes(0);
});

// it("allows the user to click a button via pressing the space key when a `LinkButton` property is initialised inside ComponentRegister without the `onKeyDown` property", () => {
//   const mockClickHandler = jest.fn();

//   const ComponentWithoutKeydown: LinkButtonComponentType = ({
//     href,
//     children
//   }: LinkButtonComponentTypeProps) => <div data-href={href}>{children}</div>;

//   const handleKeydownSpy = jest.spyOn(GovukButton.prototype, "handleKeyDown");

//   ComponentRegister.init({
//     components: {
//       LinkButton: ComponentWithoutKeydown
//     }
//   });

//   const wrapper = mount(<LinkButton href="/">Testchildren</LinkButton>);

//   const div = wrapper.find("div");

//   div.getDOMNode().addEventListener("click", mockClickHandler);

//   // 32 is the javascript event keycode for the space key
//   div.simulate("keydown", { keyCode: 32 });

//   expect(handleKeydownSpy).toHaveBeenCalledTimes(1);
//   expect(mockClickHandler).toHaveBeenCalledTimes(1);
// });
