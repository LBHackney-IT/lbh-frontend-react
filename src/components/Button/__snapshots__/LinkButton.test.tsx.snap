// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with all props 1`] = `
<a
  className="govuk-button lbh-button class1 class2 govuk-button--start"
  href="/"
  id="1234"
  onKeyDown={[Function]}
  role="button"
>
  Testchildren
  <svg
    focusable="false"
    height="19"
    role="presentation"
    viewBox="0 0 33 40"
    width="17.5"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h13l20 20-20 20H0l20-20z"
      fill="currentColor"
    />
  </svg>
</a>
`;

exports[`renders correctly without optional props 1`] = `
<a
  className="govuk-button lbh-button"
  href="/"
  onKeyDown={[Function]}
  role="button"
>
  Testchildren
</a>
`;
